biz.aQute.bndlib/src/aQute/bnd/plugin/jpms/JPMSModuleInfoPlugin.java

```diff
@@ -0,0 +1,397 @@
+		}
+
+		Parameters moduleParameters = OSGiHeader.parseHeader(moduleProperty);
+		if (moduleParameters.size() == 0)
+			return;
+
+		if (moduleParameters.size() > 1)
+			throw new IllegalArgumentException("Only one -module instruction is allowed:" + moduleParameters);
+
+		Entry<String, Attrs> moduleInstructions = moduleParameters.entrySet()
+			.iterator()
+			.next();
+
+		Packages index = new Packages();
+
+		// Index the whole class path
+		for (Jar jar : analyzer.getClasspath()) {
+			String moduleName = getModuleName(analyzer, jar);
+			String moduleVersion = jar.getModuleVersion();
+			jar.getDirectories()
+				.entrySet()
+				.stream()
+				.filter(entry -> (entry.getValue() != null) && !entry.getValue()
+					.isEmpty()
+					&& entry.getKey()
+						.length() > 0
+					&& !entry.getKey()
+						.startsWith("META-INF")
+					&& !entry.getKey()
+						.startsWith("OSGI-INF")
+					&& !entry.getKey()
+						.startsWith("OSGI-OPT")
+					&& !entry.getKey()
+						.startsWith("WEB-INF"))
+				.forEach(entry -> {
+					PackageRef ref = analyzer.getPackageRef(entry.getKey());
+					Attrs attrs = new Attrs();
+					if (moduleName != null) {
+						attrs.put(INTERNAL_MODULE_DIRECTIVE, moduleName);
+					}
+					if (moduleVersion != null) {
+						attrs.put(INTERNAL_MODULE_VERSION_DIRECTIVE, moduleVersion);
+					}
+					index.put(ref, attrs);
+				});
+		}
+
+		ModuleInfoWriter writer = nameAccessAndVersion(moduleInstructions, requireCapabilities, analyzer);
+
+		requires(moduleInstructions, analyzer, index, writer);
+		exportPackages(analyzer, writer);
+		openPackages(analyzer, writer);
+		serviceLoaderProviders(provideCapabilities, analyzer, writer);
+		serviceLoaderUses(requireCapabilities, analyzer, writer);
+		mainClass(analyzer, writer);
+
+		// TODO use annotations to store other header info???
+		// AnnotationVisitor visitAnnotation = classWriter.visitAnnotation(...,
+		// false);
+
+		ByteBufferDataOutput bbout = new ByteBufferDataOutput();
+		writer.write(bbout);
+
+		analyzer.getJar()
+			.putResource(MODULE_INFO_CLASS,
+				new EmbeddedResource(bbout.toByteArray(), analyzer.lastModified()));
+	}
+
+	private String getModuleName(Analyzer analyzer, Jar jar) throws Exception {
+		String moduleName = jar.getModuleName();
+		if (moduleName == null) {
+			if (jar.getSource() != null && jar.getSource()
+				.isDirectory()) {
+				return null;
+			}
+			moduleName = jar.getName();
+			Matcher matcher = mangledModuleName.matcher(moduleName);
+			if (matcher.matches()) {
+				moduleName = matcher.group(1);
+			}
+			if (logger.isWarnEnabled())
+				logger.warn("Using generated module name '{}' for: {}", moduleName, jar);
+		}
+		return moduleName;
+	}
+
+	private int access(Parameters requireCapabilities) {
+		return requireCapabilities.entrySet()
+			.stream()
+			.filter(entry -> Header.removeDuplicateMarker(entry.getKey())
+				.equals(EXTENDER_NAMESPACE))
+			.findAny()
+			.map(entry -> ModuleAttribute.ACC_OPEN)
```

Comment:
Use `.mapToInt` since you need an `int` and not an `Integer`.

Reply:
here Optional does not have a mapToInt.

Reply:
Move the mapToInt above findAny